test/unit/auth/tenant.spec.ts

```diff
@@ -0,0 +1,272 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as _ from 'lodash';
+import * as chai from 'chai';
+import * as sinonChai from 'sinon-chai';
+import * as chaiAsPromised from 'chai-as-promised';
+
+import {deepCopy} from '../../../src/utils/deep-copy';
+import {EmailSignInConfig} from '../../../src/auth/auth-config';
+import {
+  Tenant, TenantOptions, TenantServerResponse,
+} from '../../../src/auth/tenant';
+
+
+chai.should();
+chai.use(sinonChai);
+chai.use(chaiAsPromised);
+
+const expect = chai.expect;
+
+describe('Tenant', () => {
+  const serverRequest = {
+    name: 'projects/project1/tenants/TENANT_ID',
+    displayName: 'TENANT_DISPLAY_NAME',
+    allowPasswordSignup: true,
+    enableEmailLinkSignin: true,
+  };
+
+  const clientRequest = {
+    displayName: 'TENANT_DISPLAY_NAME',
+    emailSignInConfig: {
+      enabled: true,
+      passwordRequired: false,
+    },
+  };
+
+  const tenantOptions: TenantOptions = {
+    displayName: 'TENANT_DISPLAY_NAME',
+    emailSignInConfig: {
+      enabled: true,
+      passwordRequired: false,
+    },
+  };
+
+  describe('buildServerRequest()', () => {
+    describe('for an update request', () => {
+      it('should return the expected server request', () => {
+        const tenantOptionsClientRequest = deepCopy(clientRequest);
+        const tenantOptionsServerRequest = deepCopy(serverRequest);
+        delete tenantOptionsServerRequest.name;
+        expect(Tenant.buildServerRequest(tenantOptionsClientRequest, false))
+          .to.deep.equal(tenantOptionsServerRequest);
+      });
+
+      it('should throw on invalid EmailSignInConfig object', () => {
+        const tenantOptionsClientRequest = deepCopy(clientRequest);
+        tenantOptionsClientRequest.emailSignInConfig = null;
+        expect(() => Tenant.buildServerRequest(tenantOptionsClientRequest, false))
+          .to.throw('"EmailSignInConfig" must be a non-null object.');
+      });
+
+      it('should throw on invalid EmailSignInConfig attribute', () => {
+        const tenantOptionsClientRequest = deepCopy(clientRequest) as any;
+        tenantOptionsClientRequest.emailSignInConfig.enabled = 'invalid';
+        expect(() => {
+          Tenant.buildServerRequest(tenantOptionsClientRequest, false);
+        }).to.throw('"EmailSignInConfig.enabled" must be a boolean.');
+      });
+
+      it('should throw when type is specified', () => {
+        const tenantOptionsClientRequest: TenantOptions = deepCopy(tenantOptions);
+        tenantOptionsClientRequest.type = 'lightweight';
+        expect(() => Tenant.buildServerRequest(tenantOptionsClientRequest, false))
```

Comment:
Calls like this can be made more readable if you define a constant somewhere for `false`:

```
const createRequest = true;

Tenant.buildServerRequest(tenantOptionsClientRequest, !createRequest)
```

Reply:
done